### DISCUSSION TITLE
log : reference error, log에 찍히지 않음.

### DISCUSSION THREAD

**Discussion by catcatatt** [luftaquila/monolith#27] [Q&A]
원래 가속도계와 can 트렌시버만 연결되어있을 때는 가속도 데이터가 잘 들어왔는데(2024-07-31.log), 모든 센서를 꽂으니 ReferenceError 에러가 뜨는 것(2024-08-01.log)이 의문입니다.
raw 를 해석해보면 adc1, adc2에 연결되어 있는 linear sensor, 그리고 가속도 데이터가 들어오고 있습니다. 

그리고 처음에 모든 것을 setup하는 부분에서   SYS_LOG(LOG_INFO, ECU, ECU_BOOT);
SYS_LOG(LOG_INFO, ESP, ESP_INIT);
등등 이와 같은 로그들이 뜨지 않습니다. 

정리하자면, setup 하는 부분의 SYS_LOG 로그가 찍히지 않으며, 현재 연결된 센서들 중 가속도계와, adc1,adc2 만 reference error로 로그가 찍히고 나머지 센서는 아예 들어오지 않습니다. 

[2024-07-31.log](https://github.com/user-attachments/files/16454304/2024-07-31.log)
[2024-08-01.log](https://github.com/user-attachments/files/16454305/2024-08-01.log)

아래는 monolith 를 반영한 코드입니다. adc통신 부분과 logger.h만 afa-telemetry 에서 가져왔습니다.
https://github.com/catcatatt/monolith

**Comment by luftaquila** [comment]
확인은 좀 해 봤는데요, 우선 fork해서 변경 사항을 커밋하라고 말씀드렸던 것은 변경하신 코드가 무엇인지 diff를 보기 위함입니다.

diff가 제대로 표시되려면 파일 이름은 같고 내용이 달라야 하는데요, 내용물만 덮어쓰신게 아니라 *device/TMA-1* 경로를 *device/stm32_v2* 로 이름까지 변경하셨기 때문에 diff가 제대로 표시되지 않네요.

로컬에서 따로 클론해서 diff를 확인해봤는데, 로그 포맷이 monolith 베이스가 아니라 예전에 만든 a-fa-telemetry 규격에 더 가까운 것 같네요. 서버 코드도 수정하신 것으로 알고 있는데, 올려주신 로그가 서버 로그이니 stm32 코드가 아니라 서버 측 코드를 더 중점적으로 확인해봐야 할 것 같습니다.

누락된다고 말씀하신 로그들은 sd카드에서는 저장이 되나요?

**Reply by catcatatt** [reply]
넵 우선 로그 포맷은 a-fa-telemetry 에서 가져온 것이 맞습니다. 그리고 sd카드는 사용을 하지 않고 있습니다. 
서버 코드는 어떻게 보여드리는 것이 편하실까요?
추가로 stm32 코드상에서 로그를 보내는 쪽은 문제가 없을까요?

**Reply by luftaquila** [reply]
변경하신 코드가 올라간 장비와 서버가 있지 않는 이상 코드만 읽고 디버깅을 대신 해드리기는 어렵습니다. 추측하기로는 장비와 서버 간 로그 포맷이 맞지 않아서 사람이 읽어봤을 때는 문제가 없어 보이는 raw 값에서도 에러가 나는 것 같은데요.

1. stm32에서 sd카드나 디버그 모드의 UART 출력을 사용해서 발생하는 로그를 찍어보세요. 누락됐다고 생각하시는 데이터가 stm32 자체에서 발생하지 않는 것인지, 무선 전송 중 누락되는 것인지, 서버에서 수신은 했으나 누락하는 것인지 문제 범위를 좁혀보세요.
2. 서버 코드에 `console.log()` 군데군데 찍으면서 `ReferenceError`가 발생하는 정확한 라인과 해당 로그 타입을 올바르게 디코딩하고 있는지 여부를 확인해 보세요.
3. ESP32의 핫스팟 연결이 늦어지면 장비 내부 로그 버퍼가 꽉 차서 초반의 init 로그는 drop될 수 있습니다.
4. 코드 조금 읽어보니 tachometer(휠 스피드 센서)를 ADC로 읽고 계신 것 같은데요, 해당 센서가 아날로그 전압이 아니라 디지털 펄스 파형을 출력한다면 ADC로 읽을 게 아니라 Timer Input Capture 기능을 통해서 읽어야 합니다. (모노리스의 PULSE 입력 채널)
5. CAN 통신 전체가 관측되지 않는다면 CAN 버스 에러가 발생하고 있는 것이 아닌지 점검할 필요가 있어 보입니다. CAN to USB (Candapter)같은 장비로 컴퓨터에서 읽었을 때는 CAN 통신이 정상적으로 확인되나요?

**Comment by catcatatt** [comment]
1. 우선 디버그 모드의 UART 출력을 사용해보도록 하겠습니다.
2. 지금 현재는 아예 페이로드가 들어오고 있지 않은 상태라, 1처럼 stm32에서 먼저 로그를 찍어보려구요.
3. 로그 버퍼 크기를 늘리거나 로그가 기록될 때마다 버퍼가 꽉 차는지 체크하고, 꽉 찼다면 별도로 백업하는 동적 로그 버퍼 관리 시스템을 구현해보겠습니다.
4. 휠스피드 센서가 아날로그로 출력된다고 합니다.
5. CAN to USB 가 없어서, 이 역시 1로 디버깅해보겠습니다.

디버깅 해보고 또 문의드릴 문제가 생기면 다시 답글 달겠습니다.
감사합니다:)

**Reply by luftaquila** [reply]
올려주신 로그 파일이 서버에서 처리하는 로그가 아닌가요? 2번은 말씀하신 `ReferenceError` 가 나는 로그들에 대한 내용입니다. 들어오는 로그도 파싱이 제대로 안 되고 있으니 로그 프로토콜도 한 번 점검해 보세요.

모노리스의 텔레메트리는 측정값을 주행 중에 실시간으로 확인하기 위한 용도로, 발생하는 데이터 전체를 전송하는 것을 보장하지 않습니다. 발생하는 모든 데이터에 대한 완전한 저장을 보장하는 것은 SD카드 뿐입니다.  UART 디버그 모드 또한 UART 전송 오버헤드가 매우 큰 관계로 실제 주행 상황과 동일한 로그 발생을 보장하지 않습니다. (로그 발생 속도보다 UART 전송 속도가 압도적으로 느립니다.)

STM32가 ESP32로 전송하기 위해 사용하는 버퍼 크기는 이미 16KB로, 1024개의 로그를 저장할 수 있으며 충분히 불필요하게 거대합니다(다음 버전에서는 4KB 이하로 축소시킬 계획입니다). 불안정한 무선 환경 특성상 Wi-Fi 연결이 잠시 끊기면 전송하지 못한 로그가 모두 쌓이다가, 다시 연결되면 쌓인 버퍼를 dump하느라 실시간 데이터를 보내주지 못하는 문제가 생깁니다. 이러한 이유로 ESP32측 I2C 수신 버퍼는 1KB에 불과합니다. Bufferbloat 현상에 대해 알아보시면 버퍼 크기를 늘리는 것은 도움이 되지 않는다는 것을 알 수 있으실 겁니다. 버퍼가 꽉 찼다면 별도로 백업하는 더블 버퍼링은 일반적으로 단순히 버퍼 크기를 무한정 늘리는 것보다 더 비효율적입니다.

**Reply by catcatatt** [reply]
마지막 답변을 이제야 봤습니다.  디버그 메시지를 uart 디버그를 사용해서 아두이노 시리얼 모니터로 확인한 결과 글자가 너무 깨져서 출력되어 확인할 수 없었습니다.(이부분을 계속 수정하고자 했는데, UART 전송 오버헤드가 매우 컸던 이유같습니다), 
그래서 사용하지 않고 있던 led pin으로 디버깅을 한 결과 esp32와 handshaking 하는 부분, sys_log() 를 큐에 넣고, 큐에서 빼내어 i2c를 사용해서 esp32로 보내는 부분, adc를 통해 linear sensor 값이 배열에 담기는 부분을 확인한 결과 다 작동을 하는 것으로 보입니다. 실제로 어떤 로그가 찍히는지는 sd카드를 활용해보도록 하겠습니다.

추가로 can adapter 을 처음 사용해보는데, pc에 can to usb를 꽂았고 SavvyCAN을 사용하고 있는데 여기에 port가 연결이 안되는 문제가 발생합니다, 이건 아예 can 버스가 잘못되어서 발생하는 문젠가요?

지금 다시 monolith 반영하고 서버 수정한 부분을 아래의 링크로 올렸습니다.
https://github.com/catcatatt/monolith

**Reply by luftaquila** [reply]
글자가 깨지는건 단순히 bps가 안 맞아서 그런 것 같은데요, 115200bps로 연결하셔야 합니다. 오버헤드가 심하다는 건 UART 디버그 모드 사용 시 메시지 전송이 완료될 때까지 CPU가 놀기 때문에 그 사이에 발생하는 로그가 모두 누락된다는 의미고, 디버그 출력 자체는 정상적으로 표시되어야 합니다.

사용하시는 candapter는 제조사에서 나오는 전용 프로그램과 드라이버가 있습니다.

수정하신 코드의 디버깅까지 제가 해드릴 수는 없습니다. 문제 상황을 말씀하시면 점검할 부분이나 놓치신 부분을 말씀드릴 수는 있지만, 변경한 코드를 실행해볼 수 있는 것도 아니기 때문에 재현이 불가능한 환경에서 코드만 읽고 stm32-esp32-서버 세 군데 중 어디서 문제가 발생하는지 확인하는건 불가능합니다. 디버깅은 직접 하시고 문제 원인을 좀 더 좁힌 다음에 다시 여쭤보시면 좋을 듯 합니다.
